### PR TITLE
Implement trajectory

### PR DESCRIPTION
**Pull request by QuantumOtterDev**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/drive/Drive.java [hunk 1]
```diff
@@ -36,10 +36,10 @@
 
   private MotorType DRIVE_MOTOR_TYPE = MotorType.kBrushless;
 
-  private boolean gyroDisabled = false;
+  private boolean gyroDisabled = true;
 
   public Drive(Gyroscope gyroscope) {
-    this.leftGearbox = new Gearbox(new Encoder(0, 1), new CANSparkMax(LEFT_BACK_CAN_ID, DRIVE_MOTOR_TYPE),
+    this.leftGearbox = new Gearbox(new Encoder(7, 8, true), new CANSparkMax(LEFT_BACK_CAN_ID, DRIVE_MOTOR_TYPE),
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/drive/Drive.java]:
```suggestion
    this.leftGearbox = new Gearbox(new Encoder(7, 8), new CANSparkMax(LEFT_BACK_CAN_ID, DRIVE_MOTOR_TYPE),
```

Let's not invert encoders here. I added this for testing.

File: src/main/java/frc/robot/subsystems/drive/Drive.java [hunk 2]
```diff
@@ -146,12 +149,12 @@ public void curvatureDrive(double throttle, double curvature) {
         this.desiredAngle = gyroscope.getGyroAngle(); 
       }
       curvature = this.getAngularError(desiredAngle) * P_VALUE; 
-      this.setLeftPower(throttle - curvature);
+      this.setLeftPower((throttle - curvature));
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/drive/Drive.java]:
```suggestion
      this.setLeftPower(throttle - curvature);
```

File: src/main/java/frc/robot/subsystems/drive/Drive.java [hunk 3]
```diff
@@ -146,12 +149,12 @@ public void curvatureDrive(double throttle, double curvature) {
...
       this.setRightPower(throttle + curvature);
     }
     else { // when robot isn't driving straight
       this.desiredAngle = Integer.MAX_VALUE; //if turn is greater than 0 or if robot is still
-      differentialDrive.curvatureDrive(-throttle, curvature, (Math.abs(throttle) < QUICK_TURN_THROTTLE_DEADZONE));
+      differentialDrive.curvatureDrive(-throttle, curvature, Math.abs(throttle) < QUICK_TURN_THROTTLE_DEADZONE);
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/drive/Drive.java]:
```suggestion
      differentialDrive.curvatureDrive(-throttle, curvature, (Math.abs(throttle) < QUICK_TURN_THROTTLE_DEADZONE));
```

File: src/main/java/frc/robot/subsystems/drive/Drive.java [hunk 4]
```diff
@@ -82,7 +84,7 @@ public void resetEncoders() {
   /**
    * Returns the current wheel speeds of the robot.
    *
-   * @return The current wheel speeds.
+   * @return The current wheel speeds, in meters per second
    */
   public DifferentialDriveWheelSpeeds getWheelSpeeds() {
     return new DifferentialDriveWheelSpeeds(this.leftGearbox.getEncoderDistance(), this.rightGearbox.getEncoderDistance());
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/drive/Drive.java]:
```suggestion
    return new DifferentialDriveWheelSpeeds(this.leftGearbox.getRate(), this.rightGearbox.getRate());
```

File: src/main/java/frc/robot/subsystems/drive/Drive.java [hunk 5]
```diff
@@ -93,8 +95,9 @@ public DifferentialDriveWheelSpeeds getWheelSpeeds() {
    *
    * @param pose The pose to which to set the odometry.
    */
-  public void resetOdementry(Pose2d pose2d) {
+  public void resetOdometry(Pose2d pose2d) {
     resetEncoders();
+    this.gyroscope.resetGyro();
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/drive/Drive.java]: Let's get rid of this
```suggestion
```

File: src/main/java/frc/robot/subsystems/drive/Drive.java [hunk 6]
```diff
@@ -56,11 +56,13 @@ public Drive(Gyroscope gyroscope) {
     this.gyroscope = gyroscope;
     this.differentialDriveOdometry = new DifferentialDriveOdometry(this.gyroscope.getRotation2d());
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/drive/Drive.java:59]: We will want to do a reset encoders before this line
```suggestion
    this.resetEncoders();
    this.differentialDriveOdometry = new DifferentialDriveOdometry(this.gyroscope.getRotation2d());
```

File: src/main/java/frc/robot/subsystems/drive/Drive.java [hunk 7]
```diff
@@ -56,11 +56,13 @@ public Drive(Gyroscope gyroscope) {
...
 
-    leftGearbox.setInverted(true);
+    //leftGearbox.setInverted(true);
+    rightGearbox.setInverted(true);
   }
 
   @Override
   public void periodic() {
+    //System.out.println(this.gyroscope.getGyroAngle());
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/drive/Drive.java]:
```suggestion
```

File: src/main/java/frc/robot/subsystems/drive/Drive.java [hunk 8]
```diff
@@ -56,11 +56,13 @@ public Drive(Gyroscope gyroscope) {
     this.gyroscope = gyroscope;
     this.differentialDriveOdometry = new DifferentialDriveOdometry(this.gyroscope.getRotation2d());
 
-    leftGearbox.setInverted(true);
+    //leftGearbox.setInverted(true);
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/drive/Drive.java]:
```suggestion
```

File: src/main/java/frc/robot/constants/DriveConstants.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+    public static final double ksVolts = 0.046401;
+    public static final double kvVoltSecondsPerMeter = 0.14137;
+    public static final double kaVoltSecondsSquaredPerMeter = 0.0068514;
+    public static final DifferentialDriveKinematics kDriveKinematics = new DifferentialDriveKinematics(.68);
+    public static final double kPDriveVel = 0.27454;
+
+
+   
+
```
Comment (kushagharahi) [inline src/main/java/frc/robot/constants/DriveConstants.java]:
```suggestion
```

File: src/main/java/frc/robot/commands/autonomousCommands/TrajectoryTutCommand.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    public TrajectoryTutCommand(Drive drive) {
+         // Create a voltage constraint to ensure we don't accelerate too fast
+        this.drive = drive;
+        this.autoVoltageConstraint = new DifferentialDriveVoltageConstraint(
+            new SimpleMotorFeedforward(
+                DriveConstants.ksVolts,
+                DriveConstants.kvVoltSecondsPerMeter,
+                DriveConstants.kaVoltSecondsSquaredPerMeter),
+            DriveConstants.kDriveKinematics,
+            10);
```
Comment (EganJ) [inline src/main/java/frc/robot/commands/autonomousCommands/TrajectoryTutCommand.java]: The `10` here refers to the max voltage, but it's not obvious from reading this code. Adding a comment or declaring it as a constant variable with a helpful name could make more readable.

File: src/main/java/frc/robot/commands/autonomousCommands/TrajectoryTutCommand.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+                        DriveConstants.kaVoltSecondsSquaredPerMeter),
+                    DriveConstants.kDriveKinematics,
+                    this.drive::getWheelSpeeds,
+                    new PIDController(DriveConstants.kPDriveVel, 0, 0),
+                    new PIDController(DriveConstants.kPDriveVel, 0, 0),
+                    // RamseteCommand passes volts to the callback
+                    this.drive::tankDriveVolts,
+                    this.drive);
+                        this.drive.resetOdometry(exampleTrajectory.getInitialPose());
+                        addCommands(this.ramseteCommand);
```
Comment (EganJ) [inline src/main/java/frc/robot/commands/autonomousCommands/TrajectoryTutCommand.java]: The formatting on 78 and 79 here make it look like they are part of the parameters to the RamseteCommand constructor. Adding a newline and changing the indentation can help. Again, it's just a readability issue, but it's good to get in the habit.

File: src/main/java/frc/robot/constants/DriveConstants.java [hunk 12]
```diff
@@ -0,0 +1,16 @@
+package frc.robot.constants;
+
+import edu.wpi.first.math.kinematics.DifferentialDriveKinematics;
+
+public class DriveConstants {
+
+    public static final double ksVolts = 0.046401;
+    public static final double kvVoltSecondsPerMeter = 0.14137;
+    public static final double kaVoltSecondsSquaredPerMeter = 0.0068514;
+    public static final DifferentialDriveKinematics kDriveKinematics = new DifferentialDriveKinematics(.68);
```
Comment (EganJ) [inline src/main/java/frc/robot/constants/DriveConstants.java]:
```suggestion
    public static final double kDriveTrackWidthMeters = 0.68;
    public static final DifferentialDriveKinematics kDriveKinematics = new DifferentialDriveKinematics(kDriveTrackWidthMeters);
```
DifferentialDriveKinematics handles most of the times we would want to know the track width but its sometimes helpful to be able to access track width directly.

File: src/main/java/frc/robot/subsystems/drive/Drive.java [hunk 13]
```diff
@@ -128,11 +131,11 @@ public void setMaxOutput(double maxOutput) {
     this.differentialDrive.setMaxOutput(maxOutput);
```
Comment (EganJ) [inline src/main/java/frc/robot/subsystems/drive/Drive.java:145]: Note that this does not limit the output power if we access `leftGearbox` or `rightGearbox` directly, so anything that bypasses `differentialDrive` is not limited here.

File: src/main/java/frc/robot/subsystems/drive/Gyroscope.java [hunk 14]
```diff
@@ -20,7 +20,7 @@ public Gyroscope(){
     public double getGyroAngle(){
         double runTime = Timer.getFPGATimestamp() - startTime;
         double drift = runTime * driftPerSecond;
-        return this.ahrs.getAngle() - drift;
+        return this.ahrs.getAngle(); //- drift;
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/drive/Gyroscope.java]:
```suggestion
        return this.ahrs.getAngle();
```

File: src/main/java/frc/robot/subsystems/drive/Gearbox.java [hunk 15]
```diff
@@ -18,7 +18,7 @@ public Gearbox(Encoder encoder, CANSparkMax... controllers) {
         // TODO: these values are for a specific robot and specefic encoder
         // that may not be the same for all robots. Change them to reference
         // a value in Constants.java - Egan
-        this.encoder.setDistancePerPulse((6.0 * Math.PI) / 2048.0);
+        this.encoder.setDistancePerPulse((0.1524 * Math.PI) / 2048.0);
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/drive/Gearbox.java:20]: Let's move this `0.1524` number to `DriveConstants.java` and call the variable `WHEEL_DIAMETER_METERS`

Comment (kushagharahi) [reply src/main/java/frc/robot/subsystems/drive/Gearbox.java:20]: Also get rid of the comment from Egan after doing that :)

File: src/main/java/frc/robot/subsystems/drive/Drive.java [hunk 16]
```diff
@@ -4,9 +4,11 @@
 import com.revrobotics.CANSparkMaxLowLevel.MotorType;
 
 import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
 import edu.wpi.first.math.kinematics.DifferentialDriveOdometry;
 import edu.wpi.first.wpilibj.Encoder;
 import edu.wpi.first.wpilibj.drive.DifferentialDrive;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
```
Comment (kushagharahi) [inline src/main/java/frc/robot/subsystems/drive/Drive.java]: Is this import needed?

File: src/main/java/frc/robot/commands/autonomousCommands/TrajectoryTutCommand.java (hunk 10 above)
Comment (kushagharahi) [reply src/main/java/frc/robot/commands/autonomousCommands/TrajectoryTutCommand.java]: @Robochase21 Can you make this `10` into a variable in this command? Something like `private final double MAX_VOLTAGE`?

File: src/main/java/frc/robot/commands/autonomousCommands/TrajectoryTutCommand.java [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+
+    // @Override
+    // public boolean isFinished() {
+    //     return this.ramseteCommand.isFinished();
+    // }
+
+    // @Override
+    // public void end(boolean interrupted) {
+
+    // }
```
Comment (kushagharahi) [inline src/main/java/frc/robot/commands/autonomousCommands/TrajectoryTutCommand.java]: Let's get rid of these lines

File: src/main/java/frc/robot/commands/autonomousCommands/TrajectoryTutCommand.java [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+import edu.wpi.first.math.trajectory.TrajectoryConfig;
+import edu.wpi.first.math.trajectory.TrajectoryGenerator;
+import edu.wpi.first.math.trajectory.constraint.DifferentialDriveVoltageConstraint;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import edu.wpi.first.wpilibj2.command.RamseteCommand;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import frc.robot.constants.DriveConstants;
+import frc.robot.subsystems.drive.Drive;
+
+public class TrajectoryTutCommand extends SequentialCommandGroup {
```
Comment (kushagharahi) [inline src/main/java/frc/robot/commands/autonomousCommands/TrajectoryTutCommand.java]: Let's make this class name and file name called `TrajectoryTutorialCommandGroup`

File: src/main/java/frc/robot/RobotContainer.java [hunk 19]
```diff
@@ -20,7 +18,6 @@
   // The robot's subsystems and commands are defined here...
   private final ExampleSubsystem m_exampleSubsystem = new ExampleSubsystem();
```
Comment (kushagharahi) [inline src/main/java/frc/robot/RobotContainer.java]: Let's get rid of this and the `ExampleSubsystem` java file. Don't forget to remove the import

File: src/main/java/frc/robot/commands/autonomousCommands/TrajectoryTutCommandGroup.java [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+                    DriveConstants.kDriveKinematics,
+                    this.drive::getWheelSpeeds,
+                    new PIDController(DriveConstants.kPDriveVel, 0, 0),
+                    new PIDController(DriveConstants.kPDriveVel, 0, 0),
+                    // RamseteCommand passes volts to the callback
+                    this.drive::tankDriveVolts,
+                    this.drive);
+                    this.drive.resetOdometry(exampleTrajectory.getInitialPose());
+                    addCommands(this.ramseteCommand);
+
```
Comment (kushagharahi) [inline src/main/java/frc/robot/commands/autonomousCommands/TrajectoryTutCommandGroup.java:83]:
```suggestion
                    
                    this.drive.resetOdometry(exampleTrajectory.getInitialPose());
                    addCommands(this.ramseteCommand);
```

File: src/main/java/frc/robot/subsystems/drive/Gearbox.java (hunk 15 above)
Comment (kushagharahi) [reply src/main/java/frc/robot/subsystems/drive/Gearbox.java:20]: This thread was resolved but the change wasn't pushed up. Unresolving

File: src/main/java/frc/robot/RobotContainer.java (hunk 19 above)
Comment (QuantumOtterDev) [reply src/main/java/frc/robot/RobotContainer.java]: did you want me to git rid of ExampleCommand file also

**Review by kushagharahi — APPROVED** [review]

**Pull request merged**